### PR TITLE
fix bf16 constant accuracy

### PR DESCRIPTION
**Pull request by zhuhaozhe**
This PR aims to sort out the data type for `constant`.

The constant should be promoted to float https://github.com/pytorch/pytorch/pull/105440. So there are serval changes to do:
 - Data type propagation should propagate constant node to `float` dtype if original dtype is `bfloat16`
 - We do not need to insert `to_dtype` after the `constant` node, directly init an `fp32` constant is faster.
```
    vectorized<bfloat16> tmp(value);
    vectorized <float> tmp1 = cvt_bf16_fp32(tmp);
->
    vectorized<float> tmp(value);
```
 - move `constant` out of the list for `all operations can support bf16 without converting to fp32`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105827

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -98,6 +98,8 @@ def deduce_node_dtype_by_subgraph(self, node: torch.fx.Node):
         return dtype
 
     def deduce_node_dtype(self, node: torch.fx.Node):
+        from .cpp import DTYPE_TO_COMPUTATION_DTYPE
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:101]: It is weird `common.py` depends on something from `cpp.py`. If it is a common module, shall we move `DTYPE_TO_COMPUTATION_DTYPE` here too?

Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py]: Done, moved it to `common.py`.

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -1361,48 +1373,56 @@ def get_index():
             ) as vec_checker:
                 i32_iinfo = np.iinfo(np.int32)
                 f32_iinfo = np.finfo(np.float32)
+                reset_opt_ctx_data_type(_graph)
```
[the remark this answers is not in the document]
Comment (zhuhaozhe) [reply test/inductor/test_cpu_repro.py:1376]: Changed as adviced.

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -412,6 +395,19 @@ def get_current_node_opt_ctx() -> OptimizationContext:
     return get_opt_ctx(V.interpreter.current_node)
 
 
+def get_computation_dtype(dtype) -> torch.dtype:
```
[the remark this answers is not in the document]
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:398]: Hi, Jiong. The scaler kernel `constant` also invokes `get_computaion_dtype`.

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -2110,7 +2090,9 @@ def constant(val, dtype):
                 with RecordOptimizationContext(__name__) as node_ctx:
                     opt_ctx: OptimizationContext = node_ctx.get_opt_ctx()
                     assert opt_ctx
-                    opt_ctx.dtype = dtype
+                    # VecKernel override dtype for constant
+                    # Vectorization only support int32/fp32 now
+                    # So if dtype = int64/fp64, we will cast it to int32/fp32 if possible
```
[the remark this answers is not in the document]
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:2095]: No, the first time to initialize `opt_ctx.dtype` should be https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L2553

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**